source/extensions/health_checkers/mysql/mysql.cc

```diff
@@ -0,0 +1,323 @@
+#include "extensions/health_checkers/mysql/mysql.h"
+
+#include "envoy/event/dispatcher.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/byte_order.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HealthCheckers {
+namespace MySQLHealthChecker {
+
+MySQLHealthChecker::MySQLHealthChecker(
+    const Upstream::Cluster& cluster, const envoy::api::v2::core::HealthCheck& config,
+    const envoy::config::health_checker::mysql::v2::MySQL& mysql_config,
+    Event::Dispatcher& dispatcher, Runtime::Loader& runtime, Runtime::RandomGenerator& random)
+    : Upstream::HealthCheckerImplBase(cluster, config, dispatcher, runtime, random),
+      user_(mysql_config.user()) {}
+
+MySQLHealthChecker::MySQLActiveHealthCheckSession::~MySQLActiveHealthCheckSession() {
+  if (client_) {
+    client_->close(Network::ConnectionCloseType::NoFlush);
+  }
+}
+
+template <size_t Bytes, typename Result = uint64_t>
+static inline Result read_little_endian_int(const uint8_t* data) {
+  static_assert(Bytes <= sizeof(Result), "possible integer overflow");
+
+  Result result = 0;
+
+  // we can't assume unaligned reads are permitted so we trust the standard library to efficiently
+  // perform the copy for small-sized buffers
+  std::copy(data, data + Bytes, reinterpret_cast<uint8_t*>(&result));
+
+  return from_little_endian(result);
+}
+
+template <size_t Bytes, typename T>
+static inline void write_little_endian_int(Buffer::Instance& out, T value) {
```

Comment:
I don't know the answer to this, but between the Thrift, Mongo, and now this, is there any way to share any of this type of functionality in a common EndianBufferHelper or something? cc @zuercher

Reply:
I think if we can agree on parsing directly from the buffer (a la mongo/thrift) vs linearizing the entire request before parsing (mysql), we could come up with a helper that supports all 3 protocols. Thrift requests can be almost arbitrarily large, and depending on configuration the size may not be known at the start of the request, so it's not reasonable to linearize them.

Between the three I already see 8/16/24/32/64-bit integers (some signed, some unsigned) plus both null- and non-terminated strings, so there's a lot to go into this utility already.

Reply:
Yeah, +1. @juchem can we at least file a tech debt tracking issue on figuring this out?

Reply:
I agree it's better. I'll check Mongo/Thrift implementation to see how they do it and depending on the work I'll either file a tech debt or update the PR.